File: apps/abr/scripts/evidence/verify.ts
/**
 * ABR Insights — Evidence Verification Script
 *
 * Reads evidence/pack.json + evidence/seal.json and verifies:
 *   1. Merkle root recomputed from artifact hashes matches seal.json
 *   2. Artifact count matches
 *   3. Pack status is 'sealed'
 *   4. If EVIDENCE_SEAL_KEY is set: verifies the HMAC signature
 *
 * Exits 1 on any failure — this is a BLOCKING CI gate.
 *
 * Usage:
 *   npx tsx scripts/evidence/verify.ts
 */
import { readFileSync, existsSync } from 'node:fs'
import { join } from 'node:path'
import { createHash, createHmac } from 'node:crypto'

const ROOT = process.cwd()
const EVIDENCE_DIR = join(ROOT, 'evidence')
const PACK_PATH = join(EVIDENCE_DIR, 'pack.json')
const SEAL_PATH = join(EVIDENCE_DIR, 'seal.json')

function sha256(content: string): string {
  return createHash('sha256').update(content).digest('hex')
}

function hmacSha256(key: string, content: string): string {
  return createHmac('sha256', key).update(content).digest('hex')
}

function merkleRoot(hashes: string[]): string {
  if (hashes.length === 0) return sha256('empty')
  if (hashes.length === 1) return hashes[0]!
  const pairs: string[] = []
  for (let i = 0; i < hashes.length; i += 2) {
    const left = hashes[i]!
    const right = hashes[i + 1] ?? left
    pairs.push(sha256(left + right))
  }
  return merkleRoot(pairs)
}

interface SealFile {
  seal: string
  merkleRoot: string
  sealedAt: string
  commitSha: string
  runId: string
  artifactCount: number
}

interface PackFile {
  status: string
  createdAt: string
  commitSha: string
  runId: string
  artifacts: { sha256: string }[]
}

function fail(msg: string): never {
  console.error(`\n❌ VERIFY FAILED: ${msg}\n`)
  process.exit(1)
}

function main() {
  console.log('🔍 ABR Insights — Verifying evidence seal...\n')

  if (!existsSync(PACK_PATH)) {
    fail('evidence/pack.json not found. Run collect.ts first.')
  }
  if (!existsSync(SEAL_PATH)) {
    fail('evidence/seal.json not found. Run seal.ts first.')
  }

  let pack: PackFile
  let seal: SealFile
  try {
    pack = JSON.parse(readFileSync(PACK_PATH, 'utf-8')) as PackFile
    seal = JSON.parse(readFileSync(SEAL_PATH, 'utf-8')) as SealFile
  } catch (e) {
    fail(`Failed to parse evidence files: ${String(e)}`)
  }

  // 1 — Pack must have status 'sealed'
  if (pack.status !== 'sealed') {
    fail(`pack.json status is '${pack.status}', expected 'sealed'`)
  }
  console.log('  ✅ Pack status: sealed')

  // 2 — Artifact count must match
  const expectedCount = pack.artifacts.length
  if (seal.artifactCount !== expectedCount) {
    fail(
      `Artifact count mismatch: seal says ${seal.artifactCount}, pack has ${expectedCount}`,
    )
  }
  console.log(`  ✅ Artifact count: ${expectedCount}`)

  // 3 — Recompute Merkle root from sorted artifact hashes
  const sortedHashes = pack.artifacts.map((a) => a.sha256).sort()
  const recomputed = merkleRoot(sortedHashes)
  if (recomputed !== seal.merkleRoot) {
    fail(
      `Merkle root mismatch:\n  expected : ${seal.merkleRoot}\n  got      : ${recomputed}`,
    )
  }
  console.log(`  ✅ Merkle root verified: ${recomputed.slice(0, 16)}...`)

  // 4 — commitSha consistency
  if (pack.commitSha !== seal.commitSha) {
    fail(
      `commitSha mismatch: pack='${pack.commitSha}' seal='${seal.commitSha}'`,
    )
  }
  console.log(`  ✅ commitSha: ${pack.commitSha}`)

  // 5 — HMAC verification (if key is available)
  const sealKey = process.env.EVIDENCE_SEAL_KEY
  if (sealKey) {
    const payload = JSON.stringify({
      merkleRoot: seal.merkleRoot,
      sealedAt: seal.sealedAt,
      commitSha: seal.commitSha,
      runId: seal.runId,
      artifactCount: seal.artifactCount,
    })
    const expected = hmacSha256(sealKey, payload)
    if (expected !== seal.seal) {
      fail('HMAC signature is invalid — evidence pack has been tampered with.')
    }
    console.log('  ✅ HMAC signature verified')
  } else {
    console.log(
      '  ⚠️  EVIDENCE_SEAL_KEY not set — skipping HMAC verification (structural only)',
    )
  }

  console.log('\n✅ Evidence seal is VALID.\n')
}

main()

File: tech-repo-scaffold/django-backbone/scripts/evidence/verify.py
#!/usr/bin/env python3
# cSpell:ignore merkle digestmod
"""
ABR Evidence Verifier
=====================
Reads pack.json + seal.json and verifies the HMAC-SHA256 seal.
Exits 1 (blocking) on any verification failure — use as a CI gate.

Usage:
    python scripts/evidence/verify.py [--output <dir>]

Environment:
    EVIDENCE_SEAL_KEY   HMAC secret key used in seal.py (required)

Exit codes:
    0   Seal is valid
    1   Seal is missing, tampered, or key mismatch — CI MUST fail
"""
import argparse
import hashlib
import hmac
import json
import logging
import os
import sys
from pathlib import Path
from typing import Any

logging.basicConfig(level=logging.INFO, format="%(levelname)s  %(message)s")
log = logging.getLogger(__name__)

_FAIL_BANNER = """
╔══════════════════════════════════════════════════════════════════════╗
║  ABR EVIDENCE SEAL VERIFICATION FAILED                              ║
║  This CI run CANNOT proceed — the evidence pack is untrusted.       ║
║                                                                      ║
║  Remediation:                                                        ║
║    1. Re-run the evidence:collect → evidence:seal pipeline           ║
║    2. Verify EVIDENCE_SEAL_KEY is correctly set as a CI secret       ║
║    3. Ensure no artifact was modified after sealing                  ║
╚══════════════════════════════════════════════════════════════════════╝
"""


def recompute_seal(pack: dict[str, Any], secret_key: str) -> str:
    canonical = json.dumps(pack, sort_keys=True, separators=(",", ":"))
    return hmac.new(
        key=secret_key.encode("utf-8"),
        msg=canonical.encode("utf-8"),
        digestmod=hashlib.sha256,
    ).hexdigest()


def fail(reason: str) -> None:
    log.error("VERIFICATION FAILED: %s", reason)
    print(_FAIL_BANNER, file=sys.stderr)
    sys.exit(1)


def main() -> None:
    parser = argparse.ArgumentParser(description="Verify ABR evidence seal")
    parser.add_argument(
        "--output", default="evidence-output", help="Directory with pack.json + seal.json"
    )
    args = parser.parse_args()

    output_dir = Path(args.output).resolve()
    pack_path = output_dir / "pack.json"
    seal_path = output_dir / "seal.json"

    log.info("ABR Evidence Verifier")
    log.info("  output_dir : %s", output_dir)

    if not pack_path.exists():
        fail(f"pack.json not found at {pack_path}")

    if not seal_path.exists():
        fail(
            f"seal.json not found at {seal_path} — "
            "a pack without a seal is untrusted and MUST be rejected"
        )

    seal_key = os.environ.get("EVIDENCE_SEAL_KEY", "")
    if len(seal_key) < 32:
        fail("EVIDENCE_SEAL_KEY is not set or too short — cannot verify seal")

    pack: dict[str, Any] = json.loads(pack_path.read_text(encoding="utf-8"))
    seal: dict[str, Any] = json.loads(seal_path.read_text(encoding="utf-8"))

    expected_mac = seal.get("mac", "")
    if not expected_mac:
        fail("seal.json is missing the 'mac' field")

    actual_mac = recompute_seal(pack, seal_key)

    if not hmac.compare_digest(actual_mac, expected_mac):
        fail(
            "HMAC mismatch — pack.json has been tampered with after sealing, "
            "or the wrong EVIDENCE_SEAL_KEY was used"
        )

    # Cross-check merkle root
    pack_root = pack.get("merkle_root", "")
    seal_root = seal.get("pack_merkle_root", "")
    if pack_root != seal_root:
        fail(
            f"Merkle root mismatch: pack says {pack_root[:16]}… "
            f"but seal records {seal_root[:16]}…"
        )

    n_artifacts = pack.get("artifact_count", 0)
    log.info("  ✓  HMAC-SHA256 seal is VALID")
    log.info("  ✓  Merkle root : %s…", pack_root[:32])
    log.info("  ✓  Artifacts   : %d", n_artifacts)
    log.info("ABR evidence pack verified — CI gate PASSED")


if __name__ == "__main__":
    main()
